### PR TITLE
JobPropertyStep with OptionalJobProperty & fixed ParametersDefinitionProperty

### PR DESCRIPTION
**Pull request by jglick**
Downstream of #209 and https://github.com/jenkinsci/jenkins/pull/1888.

[Incremental diff](https://github.com/jglick/workflow-plugin/compare/JobPropertyStep...jglick:JobPropertyStep%2BOptionalJobProperty)

@reviewbybees

### CODE REVIEW

File: multibranch/src/main/java/org/jenkinsci/plugins/workflow/multibranch/JobPropertyStep.java [hunk 1]
```diff
@@ -0,0 +1,177 @@
...
+            }
+            return result;
+        }
+
+    }
+
+    @Extension public static class HideSuperfluousBranchProperties extends DescriptorVisibilityFilter {
+
+        @Override public boolean filter(Object context, Descriptor descriptor) {
+            if (context instanceof WorkflowMultiBranchProject && (descriptor.clazz == RateLimitBranchProperty.class || descriptor.clazz == BuildRetentionBranchProperty.class)) {
```
Comment (amuniz) [inline multibranch/src/main/java/org/jenkinsci/plugins/workflow/multibranch/JobPropertyStep.java]: It seems hacky? If another property like `RateLimitBranchProperty` is added for some reason, it should be excluded here too. Is not there another way to exclude them?

File: multibranch/src/test/java/org/jenkinsci/plugins/workflow/multibranch/JobPropertyStepTest.java [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+        assertEquals(-1, lr.getArtifactDaysToKeep());
+        assertEquals(3, lr.getArtifactNumToKeep());
+    }
+
+    @Issue("JENKINS-30206")
+    @Test public void useParameter() throws Exception {
+        sampleRepo.init();
+        ScriptApproval.get().approveSignature("method groovy.lang.Binding hasVariable java.lang.String"); // TODO add to generic whitelist
+        sampleRepo.write("Jenkinsfile",
+                "properties([[$class: 'ParametersDefinitionProperty', parameterDefinitions: [[$class: 'StringParameterDefinition', name: 'myparam', defaultValue: 'default value']]]])\n" +
```
Comment (amuniz) [inline multibranch/src/test/java/org/jenkinsci/plugins/workflow/multibranch/JobPropertyStepTest.java]: I know it's what we have for now, but I have to say this is the longest way to define a `key=value` pair I have ever seen in my life.

**Comment by amuniz** [comment]
The compilation errors must be due to the upstream dependency, but there are some weird test failures that I'm not sure if they are related to upstream or not. Holding the bee until you confirm they are expected (and accepted).

File: multibranch/src/main/java/org/jenkinsci/plugins/workflow/multibranch/JobPropertyStep.java (hunk 1 above)
Comment (jglick) [inline multibranch/src/main/java/org/jenkinsci/plugins/workflow/multibranch/JobPropertyStep.java]: @amuniz please comment on the upstream PR except for things specifically added in this PR.

File: multibranch/src/test/java/org/jenkinsci/plugins/workflow/multibranch/JobPropertyStepTest.java (hunk 2 above)
Comment (jglick) [inline multibranch/src/test/java/org/jenkinsci/plugins/workflow/multibranch/JobPropertyStepTest.java]: @amuniz ditto

File: multibranch/src/main/java/org/jenkinsci/plugins/workflow/multibranch/JobPropertyStep.java (hunk 1 above)
Comment (amuniz) [inline multibranch/src/main/java/org/jenkinsci/plugins/workflow/multibranch/JobPropertyStep.java]: Sorry. Done.

**Comment by jglick** [comment]
> The compilation errors must be due to the upstream dependency, but there are some weird test failures

I have not bothered to publish snapshots of upstream dependencies since currently `multibranch` is excluded from the reactor build of this repository anyway. Running tests manually.

**Comment by tfennelly** [comment]
:bee:

**Comment by jglick** [comment]
Covered by #229.

**Pull request closed without merging**